File: SampleProjects/GoT-Characters-master/Objective-C/GoT/FavouriteTableViewCell.m
//
//  FavouriteTableViewCell.m
//  GoT
//
//  Created by piotrom1 on 26/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

#import "FavouriteTableViewCell.h"
#import "UIView+Additions.h"
#import <Masonry/Masonry.h>

@interface FavouriteTableViewCell ()

@property(nonatomic, strong, nonnull) UILabel *privateTitleLabel;
@property(nonatomic, strong, nonnull) UILabel *privateAbstractLabel;
@property(nonatomic, strong, nonnull) UIButton *privateFavouriteButton;

@end

@implementation FavouriteTableViewCell

- (nonnull instancetype)initWithStyle:(UITableViewCellStyle)style
                      reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:UITableViewCellStyleSubtitle
                reuseIdentifier:reuseIdentifier];
    if (self) {
        self.imageView.image = [FavouriteTableViewCell avatarImage];
        self.privateTitleLabel = [UILabel autolayoutView];
        self.privateAbstractLabel = [UILabel autolayoutView];
        self.privateAbstractLabel.numberOfLines = 2;
        self.privateFavouriteButton = [UIButton autolayoutView];
        [self.privateFavouriteButton
         setImage:[FavouriteTableViewCell favouriteOffImage]
         forState:UIControlStateNormal];
        [self.privateFavouriteButton
         setImage:[FavouriteTableViewCell favouriteOnImage]
         forState:UIControlStateSelected];
        [self.contentView addSubview:self.privateTitleLabel];
        [self.contentView addSubview:self.privateAbstractLabel];
        [self.contentView addSubview:self.privateFavouriteButton];
        [self.privateTitleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.equalTo(self.imageView.mas_right).offset(10);
            make.right.equalTo(self.privateFavouriteButton.mas_left).offset(-10);
            make.top.equalTo(self.privateTitleLabel.superview).offset(10);
        }];
        [self.privateAbstractLabel mas_makeConstraints:^(MASConstraintMaker *make) {
            make.top.equalTo(self.privateTitleLabel.mas_bottom).offset(10);
            make.left.right.equalTo(self.privateTitleLabel);
            make.bottom.equalTo(self.contentView).offset(-10);
        }];
        [self.privateFavouriteButton
         mas_makeConstraints:^(MASConstraintMaker *make) {
             make.width.and.height.equalTo(@40);
             make.right.equalTo(self.privateFavouriteButton.superview).offset(-10);
             make.centerY.equalTo(self.privateFavouriteButton.superview);
         }];
    }
    return self;
}

+ (UIImage *)avatarImage {
    static UIImage *avatarImage;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        avatarImage = [UIImage imageNamed:@"avatar"];
    });
    return avatarImage;
}

+ (UIImage *)favouriteOnImage {
    static UIImage *image;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        image = [UIImage imageNamed:@"fav_on"];
    });
    return image;
}

+ (UIImage *)favouriteOffImage {
    static UIImage *image;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        image = [UIImage imageNamed:@"fav_off"];
    });
    return image;
}

- (void)prepareForReuse {
    [super prepareForReuse];
    self.imageView.image = [FavouriteTableViewCell avatarImage];
    self.textLabel.text = @"";
    self.detailTextLabel.text = @"";
    self.titleLabel.text = @"";
    self.abstractLabel.text = @"";
    self.privateFavouriteButton.selected = NO;
}

- (UILabel *)titleLabel {
    return self.privateTitleLabel;
}

- (UILabel *)abstractLabel {
    return self.privateAbstractLabel;
}

- (UIButton *)favouriteButton {
    return self.privateFavouriteButton;
}

@end

File: SampleProjects/GoT-Characters-master/Swift/GoT/FavouriteTableViewCell.swift
//
//  FavouriteTableViewCell.swift
//  GoT
//
//  Created by piotrom1 on 26/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

import UIKit

class FavouriteTableViewCell: UITableViewCell {
    var titleLabel: UILabel {
        return privateTitleLabel
    }
    var abstractLabel: UILabel {
        return privateAbstractLabel
    }
    var favouriteButton: UIButton {
        return privateFavouriteButton
    }

    private var privateTitleLabel: UILabel!
    private var privateAbstractLabel: UILabel!
    private var privateFavouriteButton: UIButton!

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }

    class func avatarImage() -> UIImage? {
        var avatarImage: UIImage?
        var onceToken: Int = 0
        // TODO: use a static variable (`dispatch_once` is deprecated)
    if (onceToken == 0) {
            avatarImage = UIImage(named: "avatar")
        }
    onceToken = 1
        return avatarImage
    }

    class func favouriteOnImage() -> UIImage? {
        var image: UIImage?
        var onceToken: Int = 0
        // TODO: use a static variable (`dispatch_once` is deprecated)
    if (onceToken == 0) {
            image = UIImage(named: "fav_on")
        }
    onceToken = 1
        return image
    }

    class func favouriteOffImage() -> UIImage? {
        var image: UIImage?
        var onceToken: Int = 0
        // TODO: use a static variable (`dispatch_once` is deprecated)
    if (onceToken == 0) {
            image = UIImage(named: "fav_off")
        }
    onceToken = 1
        return image
    }

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: .subtitle, reuseIdentifier: reuseIdentifier)
        
        imageView?.image = FavouriteTableViewCell.avatarImage()
        privateTitleLabel = UILabel.autolayout()
        privateAbstractLabel = UILabel.autolayout()
        privateAbstractLabel.numberOfLines = 2
        privateFavouriteButton = UIButton.autolayout()
        privateFavouriteButton.setImage(FavouriteTableViewCell.favouriteOffImage(), for: .normal)
        privateFavouriteButton.setImage(FavouriteTableViewCell.favouriteOnImage(), for: .selected)
        contentView.addSubview(privateTitleLabel)
        contentView.addSubview(privateAbstractLabel)
        contentView.addSubview(privateFavouriteButton)
        privateTitleLabel.mas_makeConstraints({ make in
            _ = make?.left.equalTo()(self.imageView?.mas_right)?.offset()(10)
            _ = make?.right.equalTo()(self.privateFavouriteButton.mas_left)?.offset()(-10)
            _ = make?.top.equalTo()(self.privateTitleLabel.superview)?.offset()(10)
        })
        privateAbstractLabel.mas_makeConstraints({ make in
            _ = make?.top.equalTo()(self.privateTitleLabel.mas_bottom)?.offset()(10)
            _ = make?.left.right().equalTo()(self.privateTitleLabel)
            _ = make?.bottom.equalTo()(self.contentView)?.offset()(-10)
        })
        privateFavouriteButton.mas_makeConstraints({ make in
            _ = make?.width.and().height().equalTo()(40)
            _ = make?.right.equalTo()(self.privateFavouriteButton.superview)?.offset()(-10)
            _ = make?.centerY.equalTo()(self.privateFavouriteButton.superview)
        })
    
    }

    override func prepareForReuse() {
        super.prepareForReuse()
        imageView?.image = FavouriteTableViewCell.avatarImage()
        textLabel?.text = ""
        detailTextLabel?.text = ""
        titleLabel.text = ""
        abstractLabel.text = ""
        privateFavouriteButton.isSelected = false
    }
}
